crates/wasi-c2/src/file.rs

```diff
@@ -0,0 +1,246 @@
+            base_caps: self.base_caps,
+            inheriting_caps: self.inheriting_caps,
+            flags: self.file.get_fdflags()?,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct FileCaps {
+    flags: u32,
+}
+
+impl FileCaps {
+    pub fn empty() -> Self {
+        FileCaps { flags: 0 }
+    }
+
+    /// Checks if `other` is a subset of those capabilties:
+    pub fn contains(&self, other: &Self) -> bool {
+        self.flags & other.flags == other.flags
+    }
+
+    pub const DATASYNC: Self = FileCaps { flags: 1 };
+    pub const READ: Self = FileCaps { flags: 2 };
+    pub const SEEK: Self = FileCaps { flags: 4 };
+    pub const FDSTAT_SET_FLAGS: Self = FileCaps { flags: 8 };
+    pub const SYNC: Self = FileCaps { flags: 16 };
+    pub const TELL: Self = FileCaps { flags: 32 };
+    pub const WRITE: Self = FileCaps { flags: 64 };
+    pub const ADVISE: Self = FileCaps { flags: 128 };
+    pub const ALLOCATE: Self = FileCaps { flags: 256 };
+    pub const FILESTAT_GET: Self = FileCaps { flags: 512 };
+    pub const FILESTAT_SET_SIZE: Self = FileCaps { flags: 1024 };
+    pub const FILESTAT_SET_TIMES: Self = FileCaps { flags: 2048 };
+}
+
+impl std::ops::BitOr for FileCaps {
+    type Output = FileCaps;
+    fn bitor(self, rhs: FileCaps) -> FileCaps {
+        FileCaps {
+            flags: self.flags | rhs.flags,
+        }
+    }
+}
+
+impl std::fmt::Display for FileCaps {
+    fn fmt(&self, _f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        todo!()
+    }
+}
+
+pub struct FdStat {
+    pub filetype: Filetype,
+    pub base_caps: FileCaps,
+    pub inheriting_caps: FileCaps,
+    pub flags: FdFlags,
+}
+
+impl WasiFile for cap_std::fs::File {
+    fn datasync(&self) -> Result<(), Error> {
+        self.sync_data()?;
+        Ok(())
+    }
+    fn sync(&self) -> Result<(), Error> {
+        self.sync_all()?;
+        Ok(())
+    }
+    fn get_filetype(&self) -> Result<Filetype, Error> {
+        let meta = self.metadata()?;
+        // cap-std's Metadata/FileType only offers booleans indicating whether a file is a directory,
+        // symlink, or regular file.
+        // Directories should be excluded by the type system.
+        if meta.is_file() {
+            Ok(Filetype::RegularFile)
+        } else {
+            Err(Error::Badf) // XXX idk what to do here
+        }
+    }
+    fn get_fdflags(&self) -> Result<FdFlags, Error> {
+        // XXX cap-std doesnt expose append, dsync, nonblock, rsync, sync
+        todo!()
+    }
+    fn get_oflags(&self) -> Result<OFlags, Error> {
+        // XXX what if it was opened append, async, nonblock...
+        let perms = self.metadata()?.permissions();
+        if perms.readonly() {
+            Ok(OFlags::RDONLY)
+        } else {
+            Ok(OFlags::RDWR)
+        }
+    }
+    fn set_oflags(&self, _flags: OFlags) -> Result<(), Error> {
+        // XXX cap-std::fs::Permissions does not export a constructor to build this out of
```

Comment:
Yeah, cap-std is following std in this respect. On Unix, `PermissionsExt` does have a `from_mode`, but I'll think about adding something portable too.

Reply:
Implemented this except for making the flags canonical, which is the sort of thing we can sort out later.

On windows, should we reject this with an error or should we silently do nothing?